File: StructuralPatterns/Composite/TypeScript/Composite.ts
import { Component } from './Component';

export class Composite implements Component {
    private readonly name: string;
    private children: Component[] = [];
    constructor(name: string) {
        this.name = name;
    }
    add(child: Component): void {
        this.children.push(child);
    }
    remove(child: Component): void {
        this.children = this.children.filter(c => c !== child);
    }
    operation(): void {
        console.log(`Composite: ${this.name}`);
        for (const child of this.children) {
            child.operation();
        }
    }
}

File: StructuralPatterns/Composite/Go/composite.go
package main

import "fmt"

type Composite struct {
	name     string
	children []Component
}

func NewComposite(name string) *Composite {
	return &Composite{name: name}
}

func (c *Composite) Add(child Component) {
	c.children = append(c.children, child)
}

func (c *Composite) Remove(child Component) {
	for i, comp := range c.children {
		if comp == child {
			c.children = append(c.children[:i], c.children[i+1:]...)
			break
		}
	}
}

func (c *Composite) Operation() {
	fmt.Printf("Composite: %s\n", c.name)
	for _, child := range c.children {
		child.Operation()
	}
}
